File: webgame/actions.js
// Acoes do Jogador — validam fase/mana/slot antes de mutar, sempre chamadas
// via TCG.acoes.* (nunca mexendo no estado direto). Espelha game/actions.py.
var TCG = window.TCG || (window.TCG = {});

TCG.AcaoInvalida = class AcaoInvalida extends Error {};

function exigirFase(game, playerId, fasesPermitidas) {
  if (game.jogadorDaVez !== playerId) throw new TCG.AcaoInvalida("Não é a vez desse jogador.");
  if (!fasesPermitidas.includes(game.fase)) {
    throw new TCG.AcaoInvalida(`Ação exige a Fase ${fasesPermitidas.join("/")}, mas o jogo está na Fase ${game.fase}.`);
  }
}

function pagarMana(game, playerId, custo) {
  const ps = game.players[playerId];
  if (ps.mana < custo) throw new TCG.AcaoInvalida(`Mana insuficiente: precisa de ${custo}, tem ${ps.mana}.`);
  ps.mana -= custo;
  game.bus.emit("manaAlterada", { playerId, delta: -custo, total: ps.mana });
}

// "Encantamentos Contínuos" (GAME_DESIGN.md) — em vez de resolver e ir pra
// Pilha de Descarte como todo Encantamento normal, ficam em campo (num slot
// de magia, igual Domínio/Maldição) enquanto o efeito passivo de +Mana por
// turno estiver ativo. Custo de Mana 0 de propósito — o "custo" real é o
// sacrifício pago no próprio efeito (descarte, vida, POW/RES). Não usa um
// campo novo no CSV: o tipo continua "Encantamento", só o NOME está nesta
// lista — mesmo padrão de tabela-por-nome já usado por SONS/regGatilhoMaldicao,
// sem mexer no schema do CSV/loader. Espelha game/actions.py.
const ENCANTAMENTOS_CONTINUOS = new Set(["Oásis do Saara", "Geleiras do Ártico", "Selva Amazônica"]);

TCG.acoes = {
  // Fase de Invocação: traz um Combatente do Panteão pro slot de Monstro.
  // `continuar` (opcional): chamada quando a ação (incluindo qualquer
  // decisão de Maldição reativa que "invocar" torne elegível) está
  // totalmente resolvida — usado pelo turno automático da IA (ai.js) pra
  // só seguir pro próximo passo depois de uma decisão do jogador humano em
  // resposta, em vez de continuar o turno inteiro por cima dela.
  invocar(game, playerId, carta, continuar = () => {}) {
    exigirFase(game, playerId, ["INVOCACAO"]);
    const lado = game.board[playerId];
    if (lado.monstro !== null) throw new TCG.AcaoInvalida("Já há um combatente ativo neste lado do tabuleiro.");
    pagarMana(game, playerId, carta.custoMana);
    TCG.Deck.tirarEspecifica(game.panteoes[playerId], carta);
    // zera flags de "já usado neste turno" ao ENTRAR em campo — sem isso,
    // um combatente que atacou, voltou ao Panteão no mesmo turno (Cânion dos
    // Ventos) e foi invocado de novo depois ficaria com atacouNesteTurno
    // travado em true pra sempre (upkeep só reresta quem já está em campo).
    carta.atacouNesteTurno = false;
    carta.habilidadeUsadaNesteTurno = false;
    TCG.Board.colocarMonstro(lado, carta);
    game.bus.emit("combatenteInvocado", { playerId, carta });
    TCG.checarFimDeJogo(game);
    TCG.ofertarMaldicoesReativas(game, { tipo: "combatenteInvocado", playerId, carta }, continuar);
  },

  // Fase Principal ou de Batalha, 1x por turno: paga o Custo de Habilidade e
  // dispara o efeito da carta. A decisão de Maldição reativa (ex.: Roubo de
  // Essência) acontece ANTES do efeito da própria Habilidade resolver.
  ativarHabilidade(game, playerId, carta, continuar = () => {}) {
    exigirFase(game, playerId, ["PRINCIPAL", "BATALHA"]);
    if (game.board[playerId].monstro !== carta) throw new TCG.AcaoInvalida("Essa carta não é o combatente ativo desse jogador.");
    if (carta.custoHabilidade == null) throw new TCG.AcaoInvalida("Essa carta não tem Habilidade ativável.");
    if (carta.habilidadeUsadaNesteTurno) throw new TCG.AcaoInvalida("Habilidade já usada neste turno.");
    pagarMana(game, playerId, carta.custoHabilidade);
    carta.habilidadeUsadaNesteTurno = true;
    game.bus.emit("habilidadeAtivada", { playerId, carta });
    TCG.ofertarMaldicoesReativas(game, { tipo: "habilidadeAtivada", playerId, carta }, () => {
      TCG.executarEfeito(game, carta.nome, playerId, carta);
      continuar();
    });
  },

  // Fase Principal: joga uma carta de campo da mão — Domínio, Encantamento ou
  // Maldição, unificados no mesmo passo (ver plano). Maldição e a excecao:
  // setar (colocar virada pra baixo) e de graca — o Custo de Mana da carta
  // so e cobrado depois, na hora de revelar/ativar (ver ativarMaldicaoSetada).
  jogarCartaDeCampo(game, playerId, carta, slot = null) {
    exigirFase(game, playerId, ["PRINCIPAL"]);
    const ps = game.players[playerId];
    if (!ps.mao.includes(carta)) throw new TCG.AcaoInvalida("Essa carta não está na mão desse jogador.");

    const lado = game.board[playerId];
    if (carta.tipo === "Domínio") {
      // Só pode haver 1 Domínio ativo NO JOGO INTEIRO — "só pode haver um
      // ativo na mesa" (GAME_DESIGN.md) é a mesa compartilhada dos dois
      // jogadores, não uma por lado. Jogar um novo destrói QUALQUER um já
      // ativo, seja de quem for (o do próprio jogador OU o do oponente).
      const anteriorProprio = TCG.Board.dominioAtivo(lado);
      if (anteriorProprio) TCG.destroyCard(game, anteriorProprio, "substituído por novo Domínio");
      const anteriorOponente = TCG.dominioAtivo(game, TCG.oponenteDe(game, playerId));
      if (anteriorOponente) TCG.destroyCard(game, anteriorOponente, "substituído por novo Domínio");
      // checa DEPOIS de liberar o próprio slot (se havia um Domínio antes) e
      // ANTES de pagar mana/tirar da mão — senão, sem slot livre, a mana
      // seria cobrada e a carta perdida da mão sem ir pra lugar nenhum.
      if (slot === null && TCG.Board.slotsLivres(lado).length === 0) {
        throw new TCG.AcaoInvalida("Não há slot de magia livre (limite de 5) pra ativar o Domínio.");
      }
      // destruir o Domínio anterior pode disparar gatilhos de terceiros
      // (ex.: Caixa de Pandora força um descarte aleatório) que, em tese,
      // poderiam pegar justo ESTA carta ainda na mão antes dela ser
      // colocada em campo — confere de novo em vez de deixar indexOf(-1)
      // corromper silenciosamente outra carta da mão via splice(-1, 1).
      if (!ps.mao.includes(carta)) {
        throw new TCG.AcaoInvalida("Essa carta foi descartada por um efeito antes de poder ser ativada.");
      }
      pagarMana(game, playerId, carta.custoMana);
      ps.mao.splice(ps.mao.indexOf(carta), 1);
      const s = TCG.Board.colocarMagia(lado, carta, slot);
      game.ultimoDominioAtivadoPor = playerId; // ver TCG.dominioParaFundo
      game.bus.emit("dominioAtivado", { playerId, carta, slot: s });
      TCG.executarEfeito(game, carta.nome, playerId, carta);
    } else if (carta.tipo === "Encantamento") {
      const continuo = ENCANTAMENTOS_CONTINUOS.has(carta.nome);
      if (continuo && slot === null && TCG.Board.slotsLivres(lado).length === 0) {
        throw new TCG.AcaoInvalida("Não há slot de magia livre (limite de 5) pra jogar este Encantamento Contínuo.");
      }
      pagarMana(game, playerId, carta.custoMana);
      ps.mao.splice(ps.mao.indexOf(carta), 1);
      game.bus.emit("encantamentoJogado", { playerId, carta });
      TCG.executarEfeito(game, carta.nome, playerId, carta);
      if (continuo) {
        TCG.Board.colocarMagia(lado, carta, slot);
      } else {
        game.descarte[playerId].push(carta);
      }
    } else if (carta.tipo === "Maldição") {
      // mesmo cuidado do Domínio: checa slot livre ANTES de tirar da mão
      // (setar é grátis, mas a carta não pode simplesmente sumir).
      if (slot === null && TCG.Board.slotsLivres(lado).length === 0) {
        throw new TCG.AcaoInvalida("Não há slot de magia livre (limite de 5) pra setar a Maldição.");
      }
      ps.mao.splice(ps.mao.indexOf(carta), 1);
      carta.faceDown = true;
      const s = TCG.Board.colocarMagia(lado, carta, slot);
      game.bus.emit("maldicaoColocada", { playerId, carta, slot: s });
    } else {
      throw new TCG.AcaoInvalida(`"${carta.nome}" não é uma carta de campo (Domínio/Encantamento/Maldição).`);
    }
  },

  // Revela e resolve uma Maldição já setada — chamada pela decisão do dono
  // dentro de TCG.ofertarMaldicoesReativas (curses.js), nunca solta. E aqui,
  // na ativação, que o Custo de Mana da carta é cobrado (setar foi de
  // graça). `evento` (opcional): o evento que motivou a oferta, repassado
  // pro efeito da carta (ver comentário no topo da seção Maldições de
  // effects.js) — fica `null` se revelada manualmente (clique direto no
  // próprio verso em campo), sem um evento associado.
  ativarMaldicaoSetada(game, playerId, carta, evento = null) {
    const lado = game.board[playerId];
    if (!lado.magia.includes(carta)) throw new TCG.AcaoInvalida("Essa Maldição não está setada nesse lado do tabuleiro.");
    pagarMana(game, playerId, carta.custoMana);
    carta.faceDown = false;
    game.bus.emit("maldicaoAtivada", { playerId, carta });
    TCG.executarEfeito(game, carta.nome, playerId, carta, evento);
    TCG.destroyCard(game, carta, "Maldição ativada");
  },

  // Fase de Batalha: confronta o combatente ativo contra o do oponente (ou
  // os Pontos de Vida dele, se o campo estiver vazio). Só uma vez por turno
  // por combatente — um segundo ataque só acontece via efeito de carta
  // (ex.: Aquiles), que causa o dano direto sem passar por esta ação.
  atacar(game, playerId, oponenteId, continuar = () => {}) {
    exigirFase(game, playerId, ["BATALHA"]);
    if (game.turno === 1) throw new TCG.AcaoInvalida("Não é possível atacar no primeiro turno.");
    const atacante = game.board[playerId].monstro;
    if (!atacante) throw new TCG.AcaoInvalida("Não há combatente ativo pra atacar.");
    if (atacante.atacouNesteTurno) throw new TCG.AcaoInvalida("Esse combatente já atacou neste turno.");
    atacante.atacouNesteTurno = true;

    // Maldições reativas elegíveis pro ataque (Escudo de Gelo Absoluto,
    // Espelho das Ilusões, Barreira de Vento Cortante, Retribuição Kármica,
    // Vínculo Sombrio) precisam decidir ANTES do dano ser calculado — pro
    // jogador local isso aparece como modal; pra IA resolve sozinho (ver o
    // listener de "selecaoPedida" pro lado não-local, em ui.js).
    const evento = { tipo: "ataqueDeclarado", atacantePlayer: playerId, atacante, defensorPlayer: oponenteId, defensor: game.board[oponenteId].monstro };
    TCG.ofertarMaldicoesReativas(game, evento, () => {
      TCG.resolverAtaque(game, playerId, atacante, oponenteId, game.board[oponenteId].monstro);
      continuar();
    });
  },

  avancarFase(game) {
    return TCG.avancarFase(game);
  },

  terminarTurno(game) {
    const inicial = game.jogadorDaVez;
    do { TCG.avancarFase(game); } while (game.jogadorDaVez === inicial && !game.fimDeJogo);
  },
};

File: game/actions.py
# -*- coding: utf-8 -*-
"""Acoes do Jogador (GAME_DESIGN.md, 'Vocabulario de Mecanicas'). Cada Acao e
um objeto com `.validar(ctrl)` e `.executar(ctrl)`, chamado sempre via
`GameController.submeter_acao(acao)` — nunca direto — pra todo mundo passar
pela mesma validacao de fase/custo/mana."""
from __future__ import annotations

from dataclasses import dataclass, field

from .components import (
    AbilityCost, AttackedThisTurn, CardInfo, Fase, FaceDown, Location,
    ManaCost, TipoEncantamento, Zona,
)
from .deck import Deck
from .events import (
    AbilityActivated, AttackDeclared, CombatantSummoned, CursePlaced,
    DeckShuffled, DomainActivated, EnchantmentPlayed, ManaChanged,
)


class AcaoInvalida(RuntimeError):
    pass


def _pagar_mana(ctrl, player_id: int, custo: int) -> None:
    ps = ctrl.players[player_id]
    if ps.mana < custo:
        raise AcaoInvalida(f"Mana insuficiente: precisa de {custo}, tem {ps.mana}.")
    ps.mana -= custo
    ctrl.bus.publish(ManaChanged(player_id=player_id, delta=-custo, total=ps.mana))


def _exigir_fase(ctrl, player_id: int, *fases: Fase) -> None:
    ts = ctrl.fase_system.estado_atual(ctrl.world)
    if ts.jogador_da_vez != player_id:
        raise AcaoInvalida("Nao e a vez desse jogador.")
    if ts.fase not in fases:
        nomes = "/".join(f.name for f in fases)
        raise AcaoInvalida(f"Acao exige a Fase {nomes}, mas o jogo esta na Fase {ts.fase.name}.")


@dataclass
class SummonAction:
    """Fase de Invocacao: traz um Combatente do Panteao pro unico slot de
    Monstro do tabuleiro. Custo: Custo de Mana da carta."""
    player_id: int
    card: int

    def executar(self, ctrl) -> None:
        _exigir_fase(ctrl, self.player_id, Fase.INVOCACAO)
        lado = ctrl.board.lado(self.player_id)
        if lado.monstro is not None:
            raise AcaoInvalida("Ja ha um combatente ativo neste lado do tabuleiro.")

        panteao = ctrl.panteoes[self.player_id]
        custo = ctrl.world.get_component(self.card, ManaCost).valor
        _pagar_mana(ctrl, self.player_id, custo)

        panteao.tirar_especifica(self.card)
        # zera "Habilidade usada neste turno" ao ENTRAR em campo — sem isso,
        # um combatente que usou a Habilidade, voltou ao Panteão no mesmo
        # turno (ex.: Cânion dos Ventos) e foi invocado de novo depois
        # ficaria com a flag travada em True pra sempre (o upkeep só reseta
        # quem já está em campo, game/phases.py UpkeepSystem).
        ability = ctrl.world.get_component(self.card, AbilityCost)
        if ability is not None:
            ability.usada_neste_turno = False
        # mesma lógica pro "já atacou neste turno" (Cânion dos Ventos): sem
        # isso, um combatente que atacou, voltou ao Panteão e foi invocado
        # de novo no MESMO turno ficaria impedido de atacar de novo. Já é
        # redundante com o reset feito em retornar_ao_panteao, mas mantido
        # aqui também por defesa (mesmo padrão do reset de AbilityCost acima).
        if ctrl.world.has_component(self.card, AttackedThisTurn):
            ctrl.world.remove_component(self.card, AttackedThisTurn)
        lado.colocar_monstro(self.card)
        loc = ctrl.world.get_component(self.card, Location)
        if loc:
            loc.zona = Zona.CAMPO_MONSTRO
        else:
            ctrl.world.add_component(self.card, Location(zona=Zona.CAMPO_MONSTRO))

        evento = CombatantSummoned(player_id=self.player_id, card=self.card)
        ctrl.bus.publish(evento)

        from .curses import ofertar_maldicoes_reativas
        ofertar_maldicoes_reativas(ctrl, evento)


@dataclass
class SwapCombatantAction:
    """Fase Principal: abre o Panteão e invoca um Combatente de lá, pagando
    o Custo de Mana normal — MESMO com um combatente já ativo em campo,
    caso em que ele é DESTRUÍDO antes (não devolvido ao Panteão: vai pra
    Pilha de Descarte e dispara qualquer gatilho "ao ser destruído", igual
    a qualquer outra destruição). Complementa a Invocação normal (Fase de
    Invocação, só serve com o slot vazio), dando ao jogador a opção de
    trocar de combatente no meio do turno pagando o preço de destruir o
    que já tinha — não há limite de vezes por turno além da Mana
    disponível."""
    player_id: int
    card: int

    def executar(self, ctrl) -> None:
        _exigir_fase(ctrl, self.player_id, Fase.PRINCIPAL)
        panteao = ctrl.panteoes[self.player_id]
        if self.card not in panteao.restantes():
            raise AcaoInvalida("Essa carta nao esta no Panteao desse jogador.")

        custo = ctrl.world.get_component(self.card, ManaCost).valor
        _pagar_mana(ctrl, self.player_id, custo)

        lado = ctrl.board.lado(self.player_id)
        anterior = lado.monstro
        if anterior is not None:
            ctrl.destruction_system.destruir(
                ctrl.world, anterior, motivo="substituído por invocação na Fase Principal")

        # destruir o combatente anterior pode disparar gatilhos de terceiros
        # (ex.: Caixa de Pandora força um descarte aleatório) que, em tese,
        # poderiam tirar justo ESTA carta do Panteão antes dela ser invocada
        # — confere de novo em vez de deixar tirar_especifica estourar.
        if self.card not in panteao.restantes():
            raise AcaoInvalida("Essa carta saiu do Panteão por um efeito antes de poder ser invocada.")
        panteao.tirar_especifica(self.card)
        # mesmo reset de flags de turno da Invocação normal (ver SummonAction
        # acima) — sem isso, um combatente que já usou a Habilidade ou já
        # atacou noutra passagem pelo campo neste turno (Cânion dos Ventos)
        # ficaria travado ao voltar.
        ability = ctrl.world.get_component(self.card, AbilityCost)
        if ability is not None:
            ability.usada_neste_turno = False
        if ctrl.world.has_component(self.card, AttackedThisTurn):
            ctrl.world.remove_component(self.card, AttackedThisTurn)
        lado.colocar_monstro(self.card)
        loc = ctrl.world.get_component(self.card, Location)
        if loc:
            loc.zona = Zona.CAMPO_MONSTRO
        else:
            ctrl.world.add_component(self.card, Location(zona=Zona.CAMPO_MONSTRO))

        evento = CombatantSummoned(player_id=self.player_id, card=self.card)
        ctrl.bus.publish(evento)

        from .curses import ofertar_maldicoes_reativas
        ofertar_maldicoes_reativas(ctrl, evento)


@dataclass
class ActivateAbilityAction:
    """Fase Tatica ou de Combate, 1x por turno: paga o Custo de Habilidade
    (losango) e dispara o efeito da carta (game/effects.py)."""
    player_id: int
    card: int

    def executar(self, ctrl) -> None:
        _exigir_fase(ctrl, self.player_id, Fase.PRINCIPAL, Fase.BATALHA)
        lado = ctrl.board.lado(self.player_id)
        if lado.monstro != self.card:
            raise AcaoInvalida("Essa carta nao e o combatente ativo desse jogador.")

        ability = ctrl.world.get_component(self.card, AbilityCost)
        if ability is None:
            raise AcaoInvalida("Essa carta nao tem Habilidade ativavel.")
        if ability.usada_neste_turno:
            raise AcaoInvalida("Habilidade ja usada neste turno.")

        _pagar_mana(ctrl, self.player_id, ability.valor)
        ability.usada_neste_turno = True
        evento = AbilityActivated(player_id=self.player_id, card=self.card)
        ctrl.bus.publish(evento)

        # a decisão de Maldição reativa (ex.: Roubo de Essência) acontece
        # ANTES do efeito da própria Habilidade resolver.
        def _continuar() -> None:
            info = ctrl.world.get_component(self.card, CardInfo)
            ctrl.effects.executar(ctrl, info.nome, self.player_id, self.card)

        from .curses import ofertar_maldicoes_reativas
        ofertar_maldicoes_reativas(ctrl, evento, _continuar)


@dataclass
class ActivateDomainAction:
    """Fase Tatica: ativa um Dominio da mao. So pode haver 1 ativo NO JOGO
    INTEIRO — "só pode haver um ativo na mesa" (GAME_DESIGN.md) e a mesa
    compartilhada dos dois jogadores, nao uma por lado. Ativar um novo
    destroi QUALQUER Dominio ja ativo, seja de quem for."""
    player_id: int
    card: int
    slot: int | None = None

    def executar(self, ctrl) -> None:
        _exigir_fase(ctrl, self.player_id, Fase.PRINCIPAL)
        ps = ctrl.players[self.player_id]
        if self.card not in ps.mao:
            raise AcaoInvalida("Essa carta nao esta na mao desse jogador.")

        custo = ctrl.world.get_component(self.card, ManaCost).valor
        _pagar_mana(ctrl, self.player_id, custo)

        lado = ctrl.board.lado(self.player_id)
        anterior = lado.dominio_ativo(ctrl.world, ctrl.dominio_cards)
        if anterior is not None:
            ctrl.destruction_system.destruir(ctrl.world, anterior, motivo="substituido por novo Domínio")
        lado_oponente = ctrl.board.lado(ctrl.oponente_de(self.player_id))
        anterior_oponente = lado_oponente.dominio_ativo(ctrl.world, ctrl.dominio_cards)
        if anterior_oponente is not None:
            ctrl.destruction_system.destruir(ctrl.world, anterior_oponente, motivo="substituido por novo Domínio")

        # checa ANTES de tirar da mao — senao, se nao houver slot livre, a
        # carta seria removida da mao sem ir pra lugar nenhum (perdida).
        if self.slot is None and not lado.slots_livres():
            raise AcaoInvalida("Nao ha slot de magia livre (limite de 5) pra ativar o Domínio.")

        # destruir o Domínio anterior pode disparar gatilhos de terceiros
        # (ex.: Caixa de Pandora força um descarte aleatorio) que, em tese,
        # poderiam pegar justo ESTA carta ainda na mao antes dela ser
        # colocada em campo — confere de novo em vez de deixar um
        # ValueError estourar.
        if self.card not in ps.mao:
            raise AcaoInvalida("Essa carta foi descartada por um efeito antes de poder ser ativada.")
        ps.mao.remove(self.card)
        slot = lado.colocar_magia(self.card, self.slot)
        ctrl.world.add_component(self.card, Location(zona=Zona.CAMPO_MAGIA, slot=slot))
        ctrl.bus.publish(DomainActivated(player_id=self.player_id, card=self.card, dominio_substituido=anterior))

        info = ctrl.world.get_component(self.card, CardInfo)
        ctrl.effects.executar(ctrl, info.nome, self.player_id, self.card)


@dataclass
class PlayEnchantmentAction:
    """Fase Tatica: joga um Encantamento da mao. Resolve na hora e vai pra
    Pilha de Descarte — exceto dois Tipo de Encantamento que ficam em campo,
    num slot de magia (GAME_DESIGN.md):
      - CONTINUO: enquanto o bônus de +Mana por turno estiver ativo.
      - EQUIPAMENTO: "vestido" no combatente ativo no momento de jogar —
        some (vai pro Descarte) quando ESSE combatente for destruído, ou
        quando o próprio Equipamento for destruído por outro efeito (essa
        segunda metade já vem de graça do próprio efeito da carta usar
        `registrar_passivo` — ver, ex., Manto da Natureza em effects.py)."""
    player_id: int
    card: int
    slot: int | None = None

    def executar(self, ctrl) -> None:
        _exigir_fase(ctrl, self.player_id, Fase.PRINCIPAL)
        ps = ctrl.players[self.player_id]
        if self.card not in ps.mao:
            raise AcaoInvalida("Essa carta nao esta na mao desse jogador.")

        info = ctrl.world.get_component(self.card, CardInfo)
        continuo = info.tipo_encantamento is TipoEncantamento.CONTINUO
        equipamento = info.tipo_encantamento is TipoEncantamento.EQUIPAMENTO
        fica_em_campo = continuo or equipamento
        lado = ctrl.board.lado(self.player_id)
        if equipamento and lado.monstro is None:
            raise AcaoInvalida("Não há combatente ativo pra equipar.")
        if fica_em_campo and self.slot is None and not lado.slots_livres():
            raise AcaoInvalida("Não há slot de magia livre (limite de 5) pra jogar este Encantamento.")

        custo = ctrl.world.get_component(self.card, ManaCost).valor
        _pagar_mana(ctrl, self.player_id, custo)
        ps.mao.remove(self.card)

        # o alvo do Equipamento e fixado AGORA (o combatente ativo no
        # instante de jogar) — antes do efeito da carta rodar e antes dela
        # ir pro slot de magia. Fica preso a essa entidade especifica
        # mesmo que ela volte ao Panteao sem ser destruida depois.
        alvo_equipado = lado.monstro if equipamento else None

        ctrl.bus.publish(EnchantmentPlayed(player_id=self.player_id, card=self.card))
        ctrl.effects.executar(ctrl, info.nome, self.player_id, self.card)

        loc = ctrl.world.get_component(self.card, Location)
        if fica_em_campo:
            slot = lado.colocar_magia(self.card, self.slot)
            loc.zona = Zona.CAMPO_MAGIA
            loc.slot = slot
            if equipamento and alvo_equipado is not None:
                from .events import CardDestroyed
                from .triggers import registrar_trigger
                registrar_trigger(
                    ctrl, CardDestroyed,
                    lambda evento, ctrl, card=self.card: ctrl.destruction_system.destruir(
                        ctrl.world, card, motivo="combatente equipado foi destruído"),
                    origem_card=self.card, owner_player_id=self.player_id,
                    condicao=lambda evento, ctrl, alvo=alvo_equipado: evento.card == alvo,
                )
        else:
            loc.zona = Zona.PILHA_DESCARTE


@dataclass
class SetCurseAction:
    """Fase Tatica: baixa uma Maldicao virada pra baixo num slot de magia.
    Setar e de graca — o Custo de Mana da carta so e cobrado na hora de
    revelar/ativar (ver ActivateSetCurseAction), nao aqui."""
    player_id: int
    card: int
    slot: int | None = None

    def executar(self, ctrl) -> None:
        _exigir_fase(ctrl, self.player_id, Fase.PRINCIPAL)
        ps = ctrl.players[self.player_id]
        if self.card not in ps.mao:
            raise AcaoInvalida("Essa carta nao esta na mao desse jogador.")

        lado = ctrl.board.lado(self.player_id)
        # checa ANTES de tirar da mao — senao, se nao houver slot livre, a
        # carta seria removida da mao sem ir pra lugar nenhum (perdida).
        if self.slot is None and not lado.slots_livres():
            raise AcaoInvalida("Nao ha slot de magia livre (limite de 5) pra setar a Maldição.")

        ps.mao.remove(self.card)
        slot = lado.colocar_magia(self.card, self.slot)
        ctrl.world.add_component(self.card, Location(zona=Zona.CAMPO_MAGIA, slot=slot))
        ctrl.world.add_component(self.card, FaceDown())
        ctrl.bus.publish(CursePlaced(player_id=self.player_id, card=self.card, slot=slot))


@dataclass
class ActivateSetCurseAction:
    """Revela e resolve uma Maldicao ja setada — SÓ chamada pela decisão do
    dono dentro de curses.ofertar_maldicoes_reativas, nunca por um clique
    livre da UI: o "a qualquer momento no turno do oponente"
    (GAME_DESIGN.md) descreve QUANDO a janela reativa pode aparecer, não
    uma permissão pra virar a carta sem o gatilho dela (`EFFECTS.
    registrar_gatilho_maldicao`) ter disparado — sem isso, dava pra
    ativar "negar o próximo ataque" fora de qualquer ataque, sem contexto
    nenhum (bug real, corrigido junto com a remoção do clique livre em
    webgame/ui.js). E aqui, na ativação, que o Custo de Mana da carta e
    cobrado (setar foi de graça). `evento` (opcional): o evento que
    motivou a oferta, repassado pro efeito da carta (ver comentário no
    topo da seção Maldições de effects.py); só fica `None` numa chamada
    direta de teste/demo fora do fluxo reativo (ex.: demo_mecanismos.py)."""
    player_id: int
    card: int
    evento: object | None = None

    def executar(self, ctrl) -> None:
        lado = ctrl.board.lado(self.player_id)
        if self.card not in lado.magia:
            raise AcaoInvalida("Essa Maldicao nao esta setada nesse lado do tabuleiro.")
        custo = ctrl.world.get_component(self.card, ManaCost).valor
        _pagar_mana(ctrl, self.player_id, custo)
        ctrl.world.remove_component(self.card, FaceDown)

        info = ctrl.world.get_component(self.card, CardInfo)
        from .events import CurseTriggered
        ctrl.bus.publish(CurseTriggered(player_id=self.player_id, card=self.card))
        ctrl.effects.executar(ctrl, info.nome, self.player_id, self.card, self.evento)

        ctrl.destruction_system.destruir(ctrl.world, self.card, motivo="Maldição ativada")


@dataclass
class DeclareAttackAction:
    """Fase de Batalha: confronta o combatente ativo contra o do oponente
    (ou os Pontos de Vida dele, se o campo estiver vazio)."""
    player_id: int
    oponente_id: int

    def executar(self, ctrl) -> None:
        _exigir_fase(ctrl, self.player_id, Fase.BATALHA)
        # GAME_DESIGN.md, 'Declarar um ataque': "Fase de Batalha, 1x/turno
        # por combatente, nunca no 1º turno da partida" — nenhuma das duas
        # restrições era verificada aqui antes desta correção (bug real: o
        # motor Python deixava atacar quantas vezes quisesse, e no turno 1).
        if ctrl.fase_system.estado_atual(ctrl.world).numero_turno == 1:
            raise AcaoInvalida("Nao e possivel atacar no primeiro turno.")
        atacante = ctrl.board.lado(self.player_id).monstro
        if atacante is None:
            raise AcaoInvalida("Nao ha combatente ativo pra atacar.")
        if ctrl.world.has_component(atacante, AttackedThisTurn):
            raise AcaoInvalida("Esse combatente ja atacou neste turno.")
        ctrl.world.add_component(atacante, AttackedThisTurn())
        defensor = ctrl.board.lado(self.oponente_id).monstro

        # Maldições reativas elegíveis pro ataque (Escudo de Gelo Absoluto,
        # Espelho das Ilusões, Barreira de Vento Cortante, Retribuição
        # Kármica, Vínculo Sombrio) precisam decidir ANTES do dano ser
        # calculado. O evento aqui é só um objeto de dados pra checar
        # elegibilidade — resolver_ataque publica o de verdade.
        evento = AttackDeclared(
            atacante_player=self.player_id, atacante_card=atacante,
            defensor_player=self.oponente_id, defensor_card=defensor,
        )

        def _continuar() -> None:
            defensor_atual = ctrl.board.lado(self.oponente_id).monstro
            ctrl.combat_system.resolver_ataque(ctrl.world, self.player_id, atacante, self.oponente_id, defensor_atual)

        from .curses import ofertar_maldicoes_reativas
        ofertar_maldicoes_reativas(ctrl, evento, _continuar)


@dataclass
class ShuffleAction:
    """Acao de embaralhar um deck (ex.: Troca Equivalente devolve cartas da
    mao pro Baralho Arcano e embaralha antes de comprar de novo)."""
    deck: Deck
    devolver: list[int] = field(default_factory=list)

    def executar(self, ctrl) -> None:
        for card in self.devolver:
            self.deck.adicionar(card)
            loc = ctrl.world.get_component(card, Location)
            if loc:
                loc.zona = Zona.BARALHO_ARCANO
        self.deck.embaralhar()
        ctrl.bus.publish(DeckShuffled(deck_nome=self.deck.nome))
